### PR TITLE
feat: Add new UnicodeUtil and corresponding tests

### PR DESCRIPTION
**Pull request by mohsaka**
Part of https://github.com/IBM/velox/pull/1836

This PR includes the addition of UnicodeUtil functions for use when writing future iceberg statistics. 

Linter warning

```
altera-id-dependent-backward-branch
Check warning: altera-id-dependent-backward-branch
backward branch (while loop) is ID-dependent due to variable reference to 'current' and may cause performance degradation
```

Should be ignored

### CODE REVIEW

File: velox/dwio/parquet/common/UnicodeUtil.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+#include <cstdint>
+#include <optional>
+#include <string>
+
+namespace facebook::velox::parquet {
+
+class UnicodeUtil {
+ public:
+  static std::string_view truncateStringMin(
```
Comment (aditi-pandit) [inline velox/dwio/parquet/common/UnicodeUtil.h]: Please can you add documentation for each of these functions what it does, what are the parameters and interpretation of the return type.

File: velox/dwio/parquet/tests/common/UnicodeUtilTest.cpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+using namespace facebook::velox::parquet;
+
+class UnicodeUtilTest : public testing::Test {};
+
+TEST_F(UnicodeUtilTest, truncateStringMin) {
+  // ASCII string.
+  std::string ascii = "Hello, world!";
+  EXPECT_EQ(
+      UnicodeUtil::truncateStringMin(ascii.c_str(), ascii.length(), 0), "");
```
Comment (aditi-pandit) [inline velox/dwio/parquet/tests/common/UnicodeUtilTest.cpp]: Why is length a parameter if we are always using .length() of the input string ? Can we just simply set that in the function logic ?

File: velox/dwio/parquet/common/UnicodeUtil.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    int32_t charLength;
+    auto codePoint =
+        utf8proc_codepoint(current, input + truncatedLength, charLength);
+    auto nextCodePoint = codePoint + 1;
+
+    // Check if the incremented code point is valid.
+    if (nextCodePoint != 0 && utf8proc_codepoint_valid(nextCodePoint)) {
+      std::string result;
+      result.reserve(truncatedLength + 4);
```
Comment (aditi-pandit) [inline velox/dwio/parquet/common/UnicodeUtil.cpp]: Is 4 the size of a code point ? If yes, then lets abstract a constant for it.

**Review by aditi-pandit — COMMENTED** [review]
Thanks @mohsaka for this code.

Please can you also check the gcc check warnings.

File: velox/dwio/parquet/tests/common/UnicodeUtilTest.cpp (hunk 2 above)
Comment (PingLiuPing) [reply velox/dwio/parquet/tests/common/UnicodeUtilTest.cpp]: It indicates `numCodePoints`.

**Comment by PingLiuPing** [comment]
@mohsaka Are you just copying my code?

**Comment by mohsaka** [comment]
> @mohsaka Are you just copying my code?

@PingLiuPing We are currently trying to merge your code into OSS main. If you would like, I can close this PR and you can open the PRs on main. Would be greatly appreciated. Please let me know. Thanks!

**Comment by aditi-pandit** [comment]
> > @mohsaka Are you just copying my code?
> 
> @PingLiuPing We are currently trying to merge your code into OSS main. If you would like, I can close this PR and you can open the PRs on main. Would be greatly appreciated. Please let me know. Thanks!

@mohsaka : Lets give Tony credits on the PRs he worked on internally. 

@PingLiuPing : We do want to merge all your code as soon as we can. If there are some specific parts that you want to collaborate on then we can do together.

File: velox/dwio/parquet/tests/common/UnicodeUtilTest.cpp (hunk 2 above)
Comment (mohsaka) [reply velox/dwio/parquet/tests/common/UnicodeUtilTest.cpp]: I think she was pertaining to ascii.length(). This is because the truncateStringMin/truncateStringMax functions take in a buffer and a length from a byte array. 

We can't use something like strlen, because unicode characters can include the a 0 byte as part of the character. 

Example usage from the entire code

```
    if constexpr (std::is_same_v<T, ByteArray>) {
      // TODO: 16 is default value. See DEFAULT_WRITE_METRICS_MODE_DEFAULT in
      // org.apache.iceberg.TableProperties. Need to support this table
      // property.
      const auto truncatedMin = UnicodeUtil::truncateStringMin(
          reinterpret_cast<const char*>(min_.ptr), min_.len, 16);
      std::string s;
```

**Comment by mohsaka** [comment]
Fixed all linter warnings except,

```
altera-id-dependent-backward-branch
Check warning: altera-id-dependent-backward-branch
backward branch (while loop) is ID-dependent due to variable reference to 'current' and may cause performance degradation
```

Should be ignored

File: velox/dwio/parquet/common/UnicodeUtil.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  /// Unicode code points. The result is a view into the original input buffer,
+  /// so the input must remain valid for the lifetime of the returned view.
+  ///
+  /// @param input Pointer to the UTF-8 encoded input string
+  /// @param inputLength Length of the input string in bytes
+  /// @param numCodePoints Maximum number of Unicode code points to include
+  /// @return A string_view containing the truncated string.
+  ///         If the input has fewer than numCodePoints code points, returns
+  ///         the entire input. The returned view points into the input buffer.
+  static std::string_view truncateStringMin(
```
Comment (aditi-pandit) [inline velox/dwio/parquet/common/UnicodeUtil.h]: Readabilty : truncateStringLower and truncateStringUpper are cleaner names I feel.

File: velox/dwio/parquet/common/UnicodeUtil.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+  // Collect the byte offset of each code point to avoid O(n^2) rescanning.
+  std::vector<size_t> codePointOffsets;
+  codePointOffsets.reserve(static_cast<size_t>(numCodePoints));
+  const char* current = input;
+  const char* truncatedEnd = input + truncatedLength;
+
+  while (current < truncatedEnd) {
+    codePointOffsets.push_back(static_cast<size_t>(current - input));
+    int32_t charLength = 1;
```
Comment (aditi-pandit) [inline velox/dwio/parquet/common/UnicodeUtil.cpp]: Nit : Rename variable to codePointSize.

File: velox/dwio/parquet/tests/common/UnicodeUtilTest.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+      UnicodeUtil::truncateStringMin(unicode.c_str(), len(unicode), 9),
+      "Hello, 世界");
+  EXPECT_EQ(
+      UnicodeUtil::truncateStringMin(unicode.c_str(), len(unicode), 10),
+      unicode);
+  EXPECT_EQ(
+      UnicodeUtil::truncateStringMin(unicode.c_str(), len(unicode), 20),
+      unicode);
+
+  // String with emoji (surrogate pairs).
```
Comment (aditi-pandit) [inline velox/dwio/parquet/tests/common/UnicodeUtilTest.cpp]: Please explain "surrogate pairs". As per search "Unicode surrogate pairs are a mechanism within the UTF-16 encoding scheme to represent characters with code points above U+FFFF—known as "supplementary characters"—using two 16-bit code units. They are used for characters that cannot fit in a single 16-bit unit, such as many emojis "

File: velox/dwio/parquet/tests/common/UnicodeUtilTest.cpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+  const std::string empty = "";
+  result = UnicodeUtil::truncateStringMax(empty.c_str(), len(empty), 0);
+  EXPECT_EQ(result, "");
+  result = UnicodeUtil::truncateStringMax(empty.c_str(), len(empty), 5);
+  EXPECT_EQ(result, "");
+
+  const std::string single = "a";
+  result = UnicodeUtil::truncateStringMax(single.c_str(), len(single), 1);
+  EXPECT_EQ(result, "a");
```
Comment (aditi-pandit) [inline velox/dwio/parquet/tests/common/UnicodeUtilTest.cpp]: Didn't entirely follow why this is not "b".

**Comment by aditi-pandit** [comment]
@jinchengchenghh, @majetideepak : Please can you take a look as well.

File: velox/dwio/parquet/common/UnicodeUtil.cpp [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+#include "velox/functions/lib/string/StringImpl.h"
+
+namespace facebook::velox::parquet {
+
+namespace {
+// Maximum number of bytes a UTF-8 encoded code point can occupy.
+constexpr int32_t kMaxUtf8BytesPerCodePoint = 4;
+} // namespace
+
+std::string_view UnicodeUtil::truncateStringMin(
```
Comment (jinchengchenghh) [inline velox/dwio/parquet/common/UnicodeUtil.cpp]: Looks like you only need to return the length

File: velox/dwio/parquet/common/UnicodeUtil.cpp [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+std::string UnicodeUtil::truncateStringMax(
+    const char* input,
+    int32_t inputLength,
+    int32_t numCodePoints) {
+  auto truncatedLength = functions::stringImpl::cappedByteLength<false>(
+      StringView(input, inputLength), numCodePoints);
+
+  if (truncatedLength == inputLength) {
+    return {input, static_cast<size_t>(inputLength)};
```
Comment (jinchengchenghh) [inline velox/dwio/parquet/common/UnicodeUtil.cpp]: This is a common case, return truncatedLength instead of string to remove the string copy, some case requires string, so return `std::variant<string_view, std::string>`

File: velox/dwio/parquet/tests/common/UnicodeUtilTest.cpp (hunk 7 above)
Comment (mohsaka) [reply velox/dwio/parquet/tests/common/UnicodeUtilTest.cpp]: Basically because `a` is already one character. 

This is a max so, if it were something like `aa` with a 1 codepoint parameter, then to make a max, we remove the right `a` and round up the first `a` to `b`. 

But since we aren't removing anything to the right of `a`, theres no rounding to be done.

File: velox/dwio/parquet/common/UnicodeUtil.cpp [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+constexpr int32_t kMaxUtf8BytesPerCodePoint = 4;
+} // namespace
+
+std::string_view UnicodeUtil::truncateStringLower(
+    const char* input,
+    int32_t inputLength,
+    int32_t numCodePoints) {
+  auto length = functions::stringImpl::cappedByteLength<false>(
+      StringView(input, inputLength), numCodePoints);
+  return {input, static_cast<size_t>(length)};
```
Comment (jinchengchenghh) [inline velox/dwio/parquet/common/UnicodeUtil.cpp]: return length

File: velox/dwio/parquet/common/UnicodeUtil.cpp (hunk 9 above)
Comment (mohsaka) [reply velox/dwio/parquet/common/UnicodeUtil.cpp]: Made this change. Thanks!

File: velox/dwio/parquet/common/UnicodeUtil.cpp (hunk 8 above)
Comment (mohsaka) [reply velox/dwio/parquet/common/UnicodeUtil.cpp]: Was a little confused at first as to why we just needed the length. But per the usage of this function

```
    if constexpr (std::is_same_v<T, ByteArray>) {
      // TODO: 16 is default value. See DEFAULT_WRITE_METRICS_MODE_DEFAULT in
      // org.apache.iceberg.TableProperties. Need to support this table
      // property.
      const auto truncatedMin = UnicodeUtil::truncateStringMin(
          reinterpret_cast<const char*>(min_.ptr), min_.len, 16);
      std::string s;
      this->plainEncode(
          ByteArray(
              truncatedMin.size(),
              reinterpret_cast<const uint8_t*>(truncatedMin.data())),
          &s);
      return s
```

We can actually just use the length and the original pointer to create the stats as we aren't doing any modifications to the original string.

File: velox/dwio/parquet/common/UnicodeUtil.cpp (hunk 10 above)
Comment (mohsaka) [reply velox/dwio/parquet/common/UnicodeUtil.cpp]: Thanks!

File: velox/dwio/parquet/common/UnicodeUtil.cpp [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+      auto bytesWritten = utf8proc_encode_char(
+          nextCodePoint, reinterpret_cast<utf8proc_uint8_t*>(buffer));
+      result.append(buffer, bytesWritten);
+      return result;
+    }
+  }
+
+  // No valid upper bound can be computed (e.g. all code points are U+10FFFF).
+  // Return the truncated string as a fallback.
+  return std::string{input, static_cast<size_t>(truncatedLength)};
```
Comment (jinchengchenghh) [inline velox/dwio/parquet/common/UnicodeUtil.cpp]: Please return string_view

File: velox/dwio/parquet/common/UnicodeUtil.cpp [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+#include <vector>
+#include "velox/external/utf8proc/utf8proc.h"
+#include "velox/functions/lib/string/StringImpl.h"
+
+namespace facebook::velox::parquet {
+
+namespace {
+// Maximum number of bytes a UTF-8 encoded code point can occupy.
+constexpr int32_t kMaxUtf8BytesPerCodePoint = 4;
```
Comment (jinchengchenghh) [inline velox/dwio/parquet/common/UnicodeUtil.cpp]: Move to the function itself, it is used only in one function

**Review by jinchengchenghh — COMMENTED** [review]
The truncate string is used only for iceberg, is it ok to place it in the common directory? CC @PingLiuPing Do we have other discussion before?

**Comment by mohsaka** [comment]
> The truncate string is used only for iceberg, is it ok to place it in the common directory? CC @PingLiuPing Do we have other discussion before?

@jinchengchenghh Thank you for the quick reviews. Addressed the comments except the refactor. Sorry, I don't have any context on any past conversations on this code. But some options, we can move it into parquet/writer/arrow since its writer specific? 

Or add an additional directory in the writer say parquet/writer/arrow/iceberg? 

Please let me know your thoughts. Thanks!

File: velox/dwio/parquet/common/UnicodeUtil.cpp [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+    auto nextCodePoint =
+        functions::stringImpl::detail::incrementCodePoint(codePoint);
+
+    if (nextCodePoint != 0) {
+      std::string result;
+      result.reserve(truncatedLength + kMaxUtf8BytesPerCodePoint);
+      result.assign(input, codePointOffsets[static_cast<size_t>(i)]);
+      char buffer[kMaxUtf8BytesPerCodePoint];
+      auto bytesWritten = utf8proc_encode_char(
```
Comment (jinchengchenghh) [inline velox/dwio/parquet/common/UnicodeUtil.cpp:81]: Why not write to result directly? buffer -> result.data(), and resize the string

File: velox/dwio/parquet/common/UnicodeUtil.cpp [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  // Collect the byte offset of each code point to avoid O(n^2) rescanning.
+  std::vector<size_t> codePointOffsets;
+  codePointOffsets.reserve(static_cast<size_t>(numCodePoints));
+  const char* current = input;
+  const char* truncatedEnd = input + truncatedLength;
+
+  while (current < truncatedEnd) {
+    codePointOffsets.push_back(static_cast<size_t>(current - input));
```
Comment (jinchengchenghh) [inline velox/dwio/parquet/common/UnicodeUtil.cpp:54]: Could you compare the two implementation https://github.com/facebookincubator/velox/pull/17388/changes/5a2a7468671a52f54338f4bc080f75810d32bbd0#diff-288ecea7abfdb29c69d6bee41029f975a0b37413bb6e99e0b47b8dc794d33d04R954

Comment (mohsaka) [reply velox/dwio/parquet/common/UnicodeUtil.cpp:54]: Implementation is different, being that the utf-8 strings needs code point validation while varbinary do not.

**Comment by bikramSingh91** [comment]
Hi Folks, adding AI generated review in case its relevant.

## Code Review

### Summary

This PR adds a `UnicodeUtil` class for Iceberg string statistics truncation. However, functionally equivalent utilities (`truncateUtf8` and `roundUpUtf8`) already exist in `velox/functions/lib/string/StringImpl.h` and are already used in the parquet writer (`velox/dwio/parquet/writer/arrow/Statistics.cpp`). There is also a correctness issue in the fallback behavior and a naming violation. Needs major revision.

### Issues Found

**Critical**

- **Duplicate functionality.** `truncateStringMinLength` is functionally equivalent to `functions::stringImpl::truncateUtf8` (`StringImpl.h:917`), and `truncateStringUpper` is functionally equivalent to `functions::stringImpl::roundUpUtf8` (`StringImpl.h:946`). Both existing functions are already used in `velox/dwio/parquet/writer/arrow/Statistics.cpp` (lines 795, 819) for the same Iceberg statistics purpose. `truncateStringMinLength` is literally a one-liner wrapper around `cappedByteLength` — the same internal call that `truncateUtf8` makes. Please use the existing utilities or justify why new ones are needed.

- `UnicodeUtil.cpp:106-120` — **Incorrect fallback when no code point can be incremented.** When all retained code points are at maximum (U+10FFFF), `truncateStringUpper` returns the truncated string as a "fallback." This is semantically wrong for an upper-bound computation — the truncated string equals the prefix, not something greater. The existing `roundUpUtf8` correctly returns `std::nullopt`, signaling no valid upper bound exists. A caller using the truncated string as an exclusive upper bound would produce incorrect range query results.

- `UnicodeUtil.cpp:66-75` — **`numCodePoints == 0` with non-empty input returns empty string instead of "no bound."** An empty string is not a valid upper bound. The existing `roundUpUtf8` returns `std::nullopt` here. The test at line 312-313 validates this incorrect behavior.

**Suggestion**

- `UnicodeUtil.h` / `UnicodeUtil.cpp` — **Class name violates Velox coding style.** The [coding style](https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md) explicitly says: "Never name a file or class `*Utils`, `*Helpers`, or `*Common`." If this class is retained, rename it to something concept-specific (e.g., `StringTruncation`, `IcebergStringBounds`).

- `UnicodeUtil.h:197` — **Return type `std::variant<std::string_view, std::string>` is unconventional.** The existing `roundUpUtf8` uses `std::optional<std::string>`, which clearly distinguishes "no valid upper bound" from a valid result. `std::variant` conflates the success and failure cases.

- `UnicodeUtil.cpp:86` — Variable `cp` is an abbreviation. Velox coding style requires descriptive names. Use `codePoint` (consistent with line 96).

- `UnicodeUtilTest.cpp:248` — Empty test fixture `UnicodeUtilTest` adds no value. Use `TEST()` instead of `TEST_F()`.

**Nitpick**

- `UnicodeUtilTest.cpp:282-286` — Comment about UTF-16 surrogate pairs is misleading. All Velox string handling is UTF-8; the surrogate pair encoding detail is irrelevant here and may confuse readers.

### Positive Observations

- Thorough test coverage with good edge cases: empty strings, emoji, surrogate boundaries, maximum code points, and mixed multi-byte sequences.
- Invalid UTF-8 handling is more explicit than the existing `roundUpUnicode` — the `if (codePoint == -1) continue;` pattern is clearer.
- Linter false positive correctly identified and noted in the PR description.

---

**Comment by mohsaka** [comment]
This is no longer needed. Covered by

https://github.com/facebookincubator/velox/pull/17388
velox/dwio/parquet/writer/arrow/StringTruncation.cpp
